File: word/src/main/java/com/dreampany/word/api/wordnik/WordnikManager.java
package com.dreampany.word.api.wordnik;

import com.dreampany.frame.util.DataUtil;
import com.wordnik.client.api.WordApi;
import com.wordnik.client.api.WordsApi;
import com.wordnik.client.model.Definition;
import com.wordnik.client.model.Example;
import com.wordnik.client.model.ExampleSearchResults;
import com.wordnik.client.model.Related;
import com.wordnik.client.model.SimpleDefinition;
import com.wordnik.client.model.SimpleExample;
import com.wordnik.client.model.TextPron;
import com.wordnik.client.model.WordObject;
import com.wordnik.client.model.WordOfTheDay;
import com.wordnik.client.model.WordSearchResult;
import com.wordnik.client.model.WordSearchResults;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import org.apache.commons.collections4.queue.CircularFifoQueue;
import timber.log.Timber;

/**
 * Created by Hawladar Roman on 9/3/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
@Singleton
public class WordnikManager {

    private final String WORDNIK_API_KEY = "api_key";
    private final String WORDNIK_API_KEY_VALUE_ROMANBJIT = "5c9a53f4c0e012d4cf5a66115420c073d7da523b9081dff1f";
    private final String WORDNIK_API_KEY_VALUE_DREAMPANY = "464b0c5a35f469103f3610840dc061f1c768aa1c223ffa447";
    private final String WORDNIK_API_KEY_VALUE_IFTENET = "a6714f04f26b9f14e29a920702e0f03dde4b84e98f94fe6fe";
    private final String[] KEYS = {WORDNIK_API_KEY_VALUE_ROMANBJIT, WORDNIK_API_KEY_VALUE_DREAMPANY, WORDNIK_API_KEY_VALUE_IFTENET};

    private final String RELATED_SYNONYM = "synonym";
    private final String RELATED_ANTONYM = "antonym";

    private final String RELATED_SYNONYM_ANTONYM = "synonym,antonym";

    private final List<WordsApi> wordsApis;
    private final List<WordApi> wordApis;

    private final CircularFifoQueue<Integer> queue;

    @Inject
    WordnikManager() {
        wordsApis = Collections.synchronizedList(new ArrayList<>());
        wordApis = Collections.synchronizedList(new ArrayList<>());
        queue = new CircularFifoQueue<>(KEYS.length);

        for (int index = 0; index < KEYS.length; index++) {
            String key = KEYS[index];
            queue.add(index);

            WordsApi wordsApi = new WordsApi();
            wordsApi.addHeader(WORDNIK_API_KEY, key);
            wordsApis.add(wordsApi);

            WordApi wordApi = new WordApi();
            wordApi.addHeader(WORDNIK_API_KEY, key);
            wordApis.add(wordApi);
        }


    }

    public WordnikWord getWordOfTheDay(String date, int limit) {
        for (WordsApi api : wordsApis) {
            try {
                WordOfTheDay wordOfTheDay = api.getWordOfTheDay(date);
                return getWord(wordOfTheDay, limit);
            } catch (Exception e) {
                Timber.e(e);
            }
        }
        return null;
    }

    public WordnikWord getWord(String word, int limit) {
        for (WordApi api : wordApis) {
            try {
                String useCanonical = "true";
                String includeSuggestions = "false";
                WordObject wordObject = api.getWord(word, useCanonical, includeSuggestions);
                return getWord(wordObject, limit);
            } catch (Exception e) {
                Timber.e(e);
            }
        }

        return null;
    }

    public List<WordnikWord> search(String query, int limit) {

        String includePartOfSpeech = null;
        String excludePartOfSpeech = null;
        String caseSensitive = "false";
        int minCorpusCount = 5;
        int maxCorpusCount = -1;
        int minDictionaryCount = 1;
        int maxDictionaryCount = -1;
        int minLength = 1;
        int maxLength = -1;
        int skip = 0;

        for (WordsApi api : wordsApis) {
            try {
                WordSearchResults results = api.searchWords(query,
                        includePartOfSpeech,
                        excludePartOfSpeech,
                        caseSensitive,
                        minCorpusCount,
                        maxCorpusCount,
                        minDictionaryCount,
                        maxDictionaryCount,
                        minLength,
                        maxLength,
                        skip,
                        limit
                );
                if (results != null) {
                    List<WordSearchResult> searches = results.getSearchResults();
                    if (!DataUtil.isEmpty(searches)) {
                        List<WordnikWord> words = new ArrayList<>(searches.size());
                        for (WordSearchResult result : searches) {
                            words.add(new WordnikWord(result.getWord().toLowerCase()));
                        }
                        return words;
                    }
                }

            } catch (Exception e) {
                Timber.e(e);
            }
        }
        return null;
    }

    private WordnikWord getWord(WordOfTheDay from, int limit) {
        WordnikWord word = new WordnikWord(from.getWord().toLowerCase());
        word.setPartOfSpeech(getPartOfSpeech(from));
        word.setPronunciation(getPronunciation(from));
        word.setDefinitions(getDefinitions(from));
        word.setExamples(getExamples(from));

        List<Related> relateds = getRelateds(from.getWord(), RELATED_SYNONYM_ANTONYM, limit);
        word.setSynonyms(getSynonyms(relateds));
        word.setAntonyms(getAntonyms(relateds));
        return word;
    }

    private WordnikWord getWord(WordObject from, int limit) {
        WordnikWord word = new WordnikWord(from.getWord().toLowerCase());

        List<Definition> definitions = getDefinitions(from.getWord(), limit);
        word.setPartOfSpeech(getPartOfSpeech(definitions));
        word.setPronunciation(getPronunciation(from));
        word.setDefinitions(getDefinitions(definitions));

        List<Example> examples = getExamples(from.getWord(), limit);
        word.setExamples(getExamples(examples));

        List<Related> relateds = getRelateds(from.getWord(), RELATED_SYNONYM_ANTONYM, limit);
        word.setSynonyms(getSynonyms(relateds));
        word.setAntonyms(getAntonyms(relateds));

        return word;
    }

    private String getPartOfSpeech(WordOfTheDay word) {
        List<SimpleDefinition> items = word.getDefinitions();
        if (!DataUtil.isEmpty(items)) {
            for (SimpleDefinition item : items) {
                if (!DataUtil.isEmpty(item.getPartOfSpeech())) {
                    return item.getPartOfSpeech();
                }
            }
        }
        return null;
    }

    private String getPartOfSpeech(List<Definition> items) {
        if (!DataUtil.isEmpty(items)) {
            for (Definition item : items) {
                if (!DataUtil.isEmpty(item.getPartOfSpeech())) {
                    return item.getPartOfSpeech();
                }
            }
        }
        return null;
    }

    private String getPronunciation(WordOfTheDay word) {
        return getPronunciation(word.getWord());
    }

    private String getPronunciation(WordObject word) {
        return getPronunciation(word.getWord());
    }

    private List<WordnikDefinition> getDefinitions(WordOfTheDay word) {
        List<SimpleDefinition> items = word.getDefinitions();
        if (!DataUtil.isEmpty(items)) {
            List<WordnikDefinition> definitions = new ArrayList<>(items.size());
            for (SimpleDefinition item : items) {
                definitions.add(new WordnikDefinition(item.getPartOfSpeech(), item.getText()));
            }
            return definitions;
        }
        return null;
    }

    private List<WordnikDefinition> getDefinitions(List<Definition> items) {
        if (!DataUtil.isEmpty(items)) {
            List<WordnikDefinition> definitions = new ArrayList<>(items.size());
            for (Definition item : items) {
                definitions.add(new WordnikDefinition(item.getPartOfSpeech(), item.getText()));
            }
            return definitions;
        }
        return null;
    }

    private List<String> getExamples(WordOfTheDay word) {
        List<SimpleExample> items = word.getExamples();
        if (!DataUtil.isEmpty(items)) {
            List<String> examples = new ArrayList<>(items.size());
            for (SimpleExample item : items) {
                examples.add(item.getText());
            }
            return examples;
        }
        return null;
    }

    private List<String> getExamples(List<Example> items) {
        if (!DataUtil.isEmpty(items)) {
            List<String> examples = new ArrayList<>(items.size());
            for (Example item : items) {
                examples.add(item.getText());
            }
            return examples;
        }
        return null;
    }

    private List<String> getSynonyms(List<Related> relateds) {
        Related related = getRelated(relateds, RELATED_SYNONYM);
        if (related != null) {
            return related.getWords();
        }
        return null;
    }

    private List<String> getAntonyms(List<Related> relateds) {
        Related related = getRelated(relateds, RELATED_ANTONYM);
        if (related != null) {
            return related.getWords();
        }
        return null;
    }


    private String getPronunciation(String word) {
        for (WordApi api : wordApis) {
            try {
                String sourceDictionary = null;
                String typeFormat = null;
                String useCanonical = "true";

                List<TextPron> pronunciations = api.getTextPronunciations(word, sourceDictionary, typeFormat, useCanonical, 3);
                if (!DataUtil.isEmpty(pronunciations)) {
                    String pronunciation = pronunciations.get(0).getRaw();
                    for (int index = 1; index < pronunciations.size(); index++) {
                        if (pronunciation.length() > pronunciations.get(index).getRaw().length()) {
                            pronunciation = pronunciations.get(index).getRaw();
                        }
                    }
                    pronunciation = pronunciation.replaceAll("(?s)<i>.*?</i>", "");
                    return pronunciation;
                }
            } catch (Exception e) {
                Timber.e(e);
            }
        }
        return null;
    }

    private List<Definition> getDefinitions(String word, int limit) {
        for (WordApi api : wordApis) {
            try {
                String partOfSpeech = null;
                String sourceDictionaries = null;
                String includeRelated = null;
                String useCanonical = "true";
                List<Definition> definitions = api.getDefinitions(word, partOfSpeech, sourceDictionaries, limit, includeRelated, useCanonical, null);
                return definitions;
            } catch (Exception e) {
                Timber.e(e);
            }
        }
        return null;
    }

    public List<Example> getExamples(String word, int limit) {
        for (WordApi api : wordApis) {
            try {
                String includeDuplicates = "true";
                String useCanonical = "true";
                int skip = 0;
                ExampleSearchResults results = api.getExamples(word, includeDuplicates, useCanonical, skip, limit);
                if (results != null) {
                    return results.getExamples();
                }
            } catch (Exception e) {
                Timber.e(e);
            }
        }
        return null;
    }

    public List<WordnikWord> query(String query, int limit) {
        for (WordsApi api : wordsApis) {
            String includePartOfSpeech = null;
            String excludePartOfSpeech = null;
            String caseSensitive = "false";
            int minCorpusCount = 5;
            int maxCorpusCount = -1;
            int minDictionaryCount = 1;
            int maxDictionaryCount = -1;
            int minLength = 1;
            int maxLength = -1;
            int skip = 0;

            try {
                WordSearchResults results = api.searchWords(query, includePartOfSpeech, excludePartOfSpeech,
                        caseSensitive, minCorpusCount, maxCorpusCount, minDictionaryCount, maxDictionaryCount, minLength, maxLength, skip, limit
                );

                if (results != null) {
                    List<WordSearchResult> searches = results.getSearchResults();
                    if (!DataUtil.isEmpty(searches)) {
                        List<WordnikWord> words = new ArrayList<>(searches.size());
                        for (WordSearchResult result : searches) {
                            words.add(new WordnikWord(result.getWord().toLowerCase()));
                        }
                        return words;
                    }
                }
            } catch (Exception e) {
                Timber.e(e);
            }
        }
        return null;
    }

    private List<Related> getRelateds(String word, String relationshipTypes, int limit) {
        for (WordApi api : wordApis) {
            try {
                String useCanonical = "true";

                List<Related> relateds = api.getRelatedWords(word, relationshipTypes, useCanonical, limit);
                return relateds;

            } catch (Exception e) {
                Timber.e(e);
            }
        }
        return null;
    }


    private Related getRelated(List<Related> relateds, String relationshipType) {
        if (!DataUtil.isEmpty(relateds)) {
            for (Related related : relateds) {
                if (relationshipType.equals(related.getRelationshipType())) {
                    return related;
                }
            }
        }
        return null;
    }
}

File: tools/src/main/kotlin/com/dreampany/tools/api/wordnik/WordnikManager.kt
package com.dreampany.tools.api.wordnik

import com.dreampany.framework.util.*
import com.dreampany.network.manager.NetworkManager
import com.dreampany.tools.api.wordnik.core.ClientException
import com.dreampany.tools.api.wordnik.misc.Constants
import com.dreampany.tools.api.wordnik.model.*
import com.google.common.collect.Maps
import org.apache.commons.collections4.queue.CircularFifoQueue
import org.apache.commons.lang3.tuple.MutablePair
import timber.log.Timber
import java.util.*
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.collections.ArrayList

/**
 * Created by roman on 2019-08-12
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class WordnikManager
@Inject constructor(
    private val network: NetworkManager
) {

    private val keys = arrayOf(
        Constants.ApiKey.WORDNIK_API_KEY_DREAM_DEBUG_1,
        Constants.ApiKey.WORDNIK_API_KEY_ROMANBJIT,
        Constants.ApiKey.WORDNIK_API_KEY_IFTENET,
        Constants.ApiKey.WORDNIK_API_KEY_DREAMPANY
    )

    private val wordsApis: MutableList<WordsApi>
    private val wordApis: MutableList<WordApi>

    private val indexQueue: CircularFifoQueue<Int>
    private val indexStatus: MutableMap<Int, MutablePair<Long, Int>>

    init {
        wordsApis = Collections.synchronizedList(ArrayList<WordsApi>())
        wordApis = Collections.synchronizedList(ArrayList<WordApi>())
        indexQueue = CircularFifoQueue(keys.size)
        indexStatus = Maps.newConcurrentMap()

        for (index in keys.indices) {
            val key = keys[index]
            indexQueue.add(index)
            wordsApis.add(WordsApi(keyOfApi = key))
            wordApis.add(WordApi(keyOfApi = key))
            indexStatus[index] = MutablePair.of(TimeUtilKt.currentMillis(), 0)
        }
        var randIndex = NumberUtil.nextRand(keys.size)
        while (randIndex > 0) {
            iterateQueue()
            randIndex--
        }
    }

    //region Public
    fun getWordOfTheDay(date: String, limit: Int): WordnikWord? {
        for (index in keys.indices) {
            val api = getWordsApi()
            try {
                val wordOfTheDay = api.getWordOfTheDay(date)
                return getWord(wordOfTheDay, limit)
            } catch (error: Throwable) {
                Timber.e(error)
                iterateQueue()
            }

        }
        return null
    }

    fun getWord(word: String, limit: Int): WordnikWord? {
        val wordObj = getWord(word)
        if (wordObj == null) return null
        return getWord(wordObj, limit)
    }
    //endregion

    //region Private
    private fun iterateQueue() {
        indexQueue.add(indexQueue.peek())
    }

    private fun adjustIndexStatus() {
        val index = indexQueue.peek()!!
        val pair = indexStatus[index]
        pair?.let {
            if (TimeUtil.isExpired(it.left, Constants.Delay.WordnikKey)) {
                it.setLeft(TimeUtilKt.currentMillis())
                it.setRight(0)
            }
            if (it.right > Constants.Limit.WORDNIK_KEY) {
                iterateQueue()
            }
            it.right++
        }
    }

    private fun getWordApi(): WordApi {
        adjustIndexStatus()
        return wordApis[indexQueue.peek()!!]
    }

    private fun getWordsApi(): WordsApi {
        adjustIndexStatus()
        return wordsApis[indexQueue.peek()!!]
    }

    private fun getWord(from: WordOfTheDay, limit: Int): WordnikWord {
        val word = WordnikWord(from.word!!.toLowerCase())
        word.partOfSpeech = getPartOfSpeech(from)
        //word.setPronunciation(getPronunciation(from));
        word.definitions = getDefinitions(from)
        word.examples = getExamples(from)

        val relateds = getRelateds(from.word, Constants.Word.SYNONYM_ANTONYM, limit)
        word.synonyms = getSynonyms(relateds)
        word.antonyms = getAntonyms(relateds)
        return word
    }

    private fun getWord(word: String): WordObject? {
        var word = word
        var index = 0
        var notFound = false
        while (index++ < keys.size) {
            val api = getWordApi()
            try {
                val useCanonical = "true"
                val includeSuggestions = "false"
                if (network.hasInternet()) {
                    val result = api.getWord(word, useCanonical, includeSuggestions)
                    return result
                }
            } catch (error: Throwable) {
                Timber.e(error)
                if (error is ClientException) {
                    if (error.toString().contains("404")) {
                        if (notFound) {
                            break
                        }
                        notFound = true
                        word = TextUtil.toTitleCase(word)
                        index--
                        continue
                    }
                }
                iterateQueue()
            }
        }
        return null
    }

    private fun getWordImpl(from: String, limit: Int): WordnikWord {
        val word = WordnikWord(from)

        val definitions = getDefinitions(from, limit)

        word.partOfSpeech = getPartOfSpeech(definitions)
        word.definitions = definitions
        word.examples = getExamplesBy(definitions)
        word.pronunciation = getPronunciation(from, limit)

        //List<String> examples = getExamples(from, limit);
        //word.setExamples(examples);
        //word.setExamples(getExamples(examples));

        val relateds = getRelateds(from, Constants.Word.SYNONYM_ANTONYM, limit)

        word.synonyms = getSynonyms(relateds)
        word.antonyms = getAntonyms(relateds)

        return word
    }

    private fun getWord(from: WordObject, limit: Int): WordnikWord? {
        var result: WordnikWord? = null
        from.word?.run {

            val word = WordnikWord(this)
            val pronunciations = getPronunciation(this, limit)
            val definitions = getDefinitions(this, limit)
            val relateds = getRelateds(this, Constants.Word.SYNONYM_ANTONYM, limit)

            word.partOfSpeech = getPartOfSpeech(definitions)
            word.pronunciation = pronunciations
            word.definitions = definitions
            word.examples = getExamplesBy(definitions)

            word.synonyms = getSynonyms(relateds)
            word.antonyms = getAntonyms(relateds)
        }
        return result
    }


    private fun getPartOfSpeech(word: WordOfTheDay): String? {
        val items = word.definitions
        if (!DataUtil.isEmpty(items)) {
/*            for (item in items) {
                *//*             if (!DataUtil.isEmpty(item.getPartOfSpeech())) {
                    return item.getPartOfSpeech();
                }*//*
            }*/
        }
        return null
    }

    private fun getPartOfSpeech(items: List<Definition>?): String? {
        var partOfSpeech: String? = null
        items?.forEach { def ->
            if (!def.partOfSpeech.isNullOrEmpty())
                partOfSpeech = def.partOfSpeech
            return@forEach
        }
        return partOfSpeech
    }

    private fun getDefinitions(word: WordOfTheDay): List<Definition>? {
        val items = word.definitions
        if (!DataUtil.isEmpty(items)) {
            val definitions = ArrayList<Definition>(items!!.size)
/*            for (item in items) {
                //definitions.add(new WordnikDefinition(item.getPartOfSpeech(), item.getText()));
            }*/
            return definitions
        }
        return null
    }

/*    private fun getDefinitions(items: List<Definition>?): List<Definition>? {
        if (!DataUtil.isEmpty(items)) {
            val definitions = ArrayList<Definition>(items!!.size)
            for (item in items) {
                definitions.add(Definition(partOfSpeech = item.partOfSpeech, text = item.text))
            }
            return definitions
        }
        return null
    }*/

    private fun getExamples(word: WordOfTheDay): List<String>? {
        val items = word.examples
        items?.run {
            val examples = ArrayList<String>(size)
            forEach { item ->
                examples.add(item)
            }
            return examples
        }
        return null;
    }

    private fun getExamples(items: List<Example>?): List<String>? {
        items?.run {
            val examples = ArrayList<String>(items.size)
            forEach {
                it.text?.run {
                    examples.add(this)
                }

            }
            return examples
        }
        return null
    }

    private fun getExamplesBy(definitions: List<Definition>?): List<String>? {
        if (!DataUtil.isEmpty(definitions)) {
            val examples = ArrayList<String>()
            for (def in definitions!!) {
                if (DataUtil.isEmpty(def.exampleUses))
                    continue
                def.exampleUses?.forEach {
                    it.text?.run {
                        examples.add(this)
                    }
                }
            }
            return examples
        }
        return null
    }

    private fun getSynonyms(relateds: List<Related>?): List<String>? {
        val related = getRelated(relateds, Constants.Word.SYNONYM)
        return related?.words?.toList()
    }

    private fun getAntonyms(relateds: List<Related>?): List<String>? {
        val related = getRelated(relateds, Constants.Word.ANTONYM)
        return related?.words?.toList()
    }


    private fun getPronunciation(word: String, limit: Int): String? {
        var word = word
        var index = 0
        var notFound = false
        while (index++ < keys.size) {
            val api = getWordApi()
            try {
                val sourceDictionary: String? = null
                val typeFormat: String? = null
                val useCanonical = "true"

                if (network.hasInternet()) {
                    val result = api.getTextPronunciations(
                        word,
                        useCanonical,
                        sourceDictionary,
                        typeFormat,
                        limit
                    )

                    if (!result.isNullOrEmpty()) {
                        var pronunciation = result[0].raw
                        for (ind in 1 until result.size) {
                            if (pronunciation!!.length > result[ind].raw!!.length) {
                                pronunciation = result[ind].raw
                            }
                        }
                        pronunciation = pronunciation!!.replace("(?s)<i>.*?</i>".toRegex(), "")
                        return pronunciation
                    }
                }
            } catch (error: Throwable) {
                Timber.e(error)
                if (error is ClientException) {
                    if (error.toString().contains(Constants.ResponseCode.NOT_FOUND.toString())) {
                        if (notFound) {
                            break
                        }
                        notFound = true
                        word = TextUtil.toTitleCase(word)
                        index--
                        continue
                    }
                }
                iterateQueue()
            }

        }
        return null
    }

    private fun getDefinitions(word: String, limit: Int): List<Definition>? {
        var word = word
        var index = 0
        var notFound = false
        while (index++ < keys.size) {
            val api = getWordApi()
            try {
                val partOfSpeech: String? = null
                val includeRelated = "false"
                val sourceDictionaries: Array<String>? = null
                val useCanonical = "true"
                val includeTags = "false"
                if (network.hasInternet()) {
                    val definitions = api.getDefinitions(
                        word,
                        limit,
                        partOfSpeech,
                        includeRelated,
                        sourceDictionaries,
                        useCanonical,
                        includeTags
                    )
                    return definitions.toList()
                }
            } catch (error: Exception) {
                Timber.e(error)
                if (error is ClientException) {
                    if (error.toString().contains(Constants.ResponseCode.NOT_FOUND.toString())) {
                        if (notFound) {
                            break
                        }
                        notFound = true
                        word = TextUtil.toTitleCase(word)
                        index--
                        continue
                    }
                }
                iterateQueue()
            }
        }
        return null
    }

    private fun getRelateds(word: String, relationshipTypes: String, limit: Int): List<Related>? {
        var word = word
        var index = 0
        var notFound = false
        while (index++ < keys.size) {
            val api = getWordApi()
            try {
                val useCanonical = "true"
                if (network.hasInternet()) {
                    val relateds = api.getRelatedWords(word, useCanonical, relationshipTypes, limit)
                    return relateds.toList()
                }

            } catch (error: Throwable) {
                Timber.e(error)
                if (error is ClientException) {
                    if (error.toString().contains(Constants.ResponseCode.NOT_FOUND.toString())) {
                        if (notFound) {
                            break
                        }
                        notFound = true
                        word = TextUtil.toTitleCase(word)
                        index--
                        continue
                    }
                }
                iterateQueue()
            }
        }
        return null
    }


    private fun getRelated(relateds: List<Related>?, relationshipType: String): Related? {
        var related: Related? = null
        relateds?.forEach { rel ->
            if (relationshipType == rel.relationshipType) {
                related = rel
                return@forEach
            }
        }
        return related
    }
    //endregion

    //region Blocked
    /*   fun getExamples(word: String, limit: Int): List<String>? {
        for (index in keys.indices) {
            val api = getWordApi()
            try {
                val includeDuplicates = "true"
                val useCanonical = "true"
                val skip = 0
                val results = api.getExamples(word, includeDuplicates, useCanonical, skip, limit)
                return results.examples?.toList() //Arrays.asList(results.getExamples())
            } catch (e: Exception) {
                Timber.e(e)
                iterateQueue()
            }

        }
        return null
    }

    fun query(query: String, limit: Int): List<WordnikWord>? {

        val includePartOfSpeech: String? = null
        val excludePartOfSpeech: String? = null
        val caseSensitive = "false"
        val minCorpusCount = 5
        val maxCorpusCount = -1
        val minDictionaryCount = 1
        val maxDictionaryCount = -1
        val minLength = 1
        val maxLength = -1
        val skip = 0

        for (index in keys.indices) {
            val api = getWordsApi()

            *//*            try {
                WordSearchResults results = api.searchWords(query, includePartOfSpeech, excludePartOfSpeech,
                        caseSensitive, minCorpusCount, maxCorpusCount, minDictionaryCount, maxDictionaryCount, minLength, maxLength, skip, limit
                );

                if (results != null) {
                    List<WordSearchResult> searches = results.getSearchResults();
                    if (!DataUtil.isEmpty(searches)) {
                        List<WordnikWord> words = new ArrayList<>(searches.size());
                        for (WordSearchResult result : searches) {
                            words.add(new WordnikWord(result.getWord().toLowerCase()));
                        }
                        return words;
                    }
                }
            } catch (Exception e) {
                Timber.e(e);
                iterateQueue();
            }*//*
        }
        return null
    }

        fun search(query: String, limit: Int): List<WordnikWord>? {

        val includePartOfSpeech: String? = null
        val excludePartOfSpeech: String? = null
        val caseSensitive = "false"
        val minCorpusCount = 5
        val maxCorpusCount = -1
        val minDictionaryCount = 1
        val maxDictionaryCount = -1
        val minLength = 1
        val maxLength = -1
        val skip = 0

        for (index in keys.indices) {
            val api = getWordsApi()
            /*            try {
                WordSearchResults results = api.searchWords(query,
                        includePartOfSpeech,
                        excludePartOfSpeech,
                        caseSensitive,
                        minCorpusCount,
                        maxCorpusCount,
                        minDictionaryCount,
                        maxDictionaryCount,
                        minLength,
                        maxLength,
                        skip,
                        limit
                );
                if (results != null) {
                    List<WordSearchResult> searches = results.getSearchResults();
                    if (!DataUtil.isEmpty(searches)) {
                        List<WordnikWord> words = new ArrayList<>(searches.size());
                        for (WordSearchResult result : searches) {
                            words.add(new WordnikWord(result.getWord().toLowerCase()));
                        }
                        return words;
                    }
                }

            } catch (Exception e) {
                Timber.e(e);
                iterateQueue();
            }*/
        }
        return null
    }
    */
    //endregion
}